Issue template and FAQ - autoscaler upstream sync (#188)
gardener/autoscaler: .github/ISSUE_TEMPLATE/sync_request.md

---
name: Upstream Autoscaler Sync 
about: Sync with Upstream Autoscaler
labels: kind/enhancement
title: Issue for - Sync with Upstream v1.2x
---

## What would you like to be added

> Gardener autoscaler should be synced with [Kubernetes Autoscaler 1.2x](https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.2x.y)

Reviewer: The link doesn't work , could we remove that . Also the minor release should be more generalized, please also fix the other occurences.
```suggestion
Gardener autoscaler should be synced with **Kubernetes Autoscaler 1.X.0**
```

Author: fixed link and versioning convention

## Why is this needed

* To keep the fork in sync with upstream.
* To ensure the shoot control planes are using respective version of CA for a given K8S version

## Steps

NOTE: replace `1.2x.0` with the specific autoscaler upstream version that you are trying to sync to. Replace `user` with your github user.

> - [ ] Fork `github.com/gardener/autoscaler`  into your github account say `user`. 

Reviewer: we should forward the user to the sync steps in [FAQ](https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/FAQ.md#how-do-i-sync-gardener-autoscaler-with-an-upstream-autoscaler-minor-release) and update the FAQ steps with what you have added here.
That way there is just one point of update/reference everytime.

Author: moved steps to FAQ and linked issue to FAQ

- [ ] Check out `github.com/gardener/autoscaler` under `$GOPATH/src/k8s.io`

> - [ ] Change to checked-out `autoscaler` dir and set upstream, origin and elankath (fork)

Reviewer: ```suggestion
- [ ] Change to checked-out `autoscaler` dir and set upstream, origin and user (fork)
```

Author: fixed fork name

- [ ] Go to [autoscaler releases](https://github.com/kubernetes/autoscaler/releases) and get the commit id for release `1.2x.0` . Note down the commitId.
- [ ] `git reset --hard commitId`
- [ ] `git checkout machine-controller-manager-provider`
- [ ] `git pull origin machine-controller-manager-provider`
- [ ] `git checkout -b sync-upstream-v1.2x.0`
- [ ] `git merge upstream-release-1.2x`
	- [ ] This is where changes of master get merged in. advantage of merging is it won’t change the commit hashes of already existing commits. 
	- [ ] Accept all vpa changes.
	- [ ] Accept our changes in `go.mod` and `go.sum` and `vendor` directory.  (The `hack/update-vendor.sh` script below is expected to fix things)
- [ ] Upgrade versions of `machine-controller-manager`, `machine-controller-manager-provider-aws`  and `machine-controller-manager-provider-azure` to the latest release in `go.mod`
- [ ] Run update vendor script after changing to `cluster-autoscaler` directory:  `./hack/update-vendor.sh 1.2x.0`
	- [ ] If the above still gives test issues then use `rsync` or `diff -rq` to figure out differences in `vendor` directory and synchronize them manually.

> - [ ] Create a new file `SYNC_CHANGES-1.2x.md`  summarily describing the changes done. Follow template. Use upstream release notes as a guide.

Reviewer: kindly add the location of the folder where to add this file.

Author: added folder name for sync changes

- [ ]  Run Tests
	- [ ] `cd cluster-autoscaler`
	- [ ] `go test $(go list ./... | grep -v cloudprovider | grep -v vendor | grep -v integration)` # core autoscaler unit-tests

> 	- [ ] `go test $(go list ./cloudprovider/mcm/... | grep -v vendor)` - run MCM cloud provider test

Reviewer: ```suggestion
	- [ ] `go test $(go list ./cloudprovider/mcm/... | grep -v vendor)` # runs MCM cloud provider implementation tests
```

Author: fixed step comment

> 	- [ ] `go test -race $(go list ./... | grep -v /vendor/ | grep -v vertical-pod-autoscaler/e2e | grep -v cluster-autoscaler/integration)`

Reviewer: do we want to run these tests ? I think for this the pwd should should `../cluster-autoscaler` as we are currently in the `cluster-autoscaler` directory.

Author: These were copied from: `hack/for-go-proj.sh`. I guess we can omit the last series of race tests as they test outside the autoscaler

Author: removed this test

	- [ ] Verify that binary can be created using: `../.ci/build`
	- [ ] Image can be created using the `Dockerfile` :
	- [ ] Run IT test locallys. 
		- [ ] Follow instructions at: https://github.com/gardener/autoscaler/blob/machine-controller-manager-provider/cluster-autoscaler/integration/usage.md  
		- [ ] Before running `make download-kubeconfigs`, create a folder `mkdir -p dev/kubeconfigs`
		- [ ] This target will print out a list of shell variable `EXPORT` statements. Copy paste thema and execute them
		- [ ] Now run `make test-integration`

> - [ ] Update the Availability Matrix 

Reviewer: kindly mention the location of the same using link

Author: added link to release matrix in faq step

- [ ] Update any RBAC rules if required
- [ ] Now make a commit.
- [ ] Push the sync branch to fork repo: `git push user sync-upstream-v1.2x.0`
- [ ] Create a PR.
- [ ] Add the `ok-to-test` label to the PR
- [ ] Ensure that integrationt test passes on CICD.

> - [ ] Give PR for review.

Reviewer: these are be trivial steps, no need to mention them. Only the following step is relevant.
```
- [ ] Push the sync branch to fork repo: `git push user sync-upstream-v1.2x.0`
- [ ] Create a PR on `machine-controller-manager-provider` branch of `gardener/autoscaler`
```

Author: adopted suggested changes

---

## What would you like to be added

> Gardener autoscaler should be synced with respecrtive [Kubernetes Autoscaler 1.x.y Release](https://github.com/kubernetes/autoscaler/releases/)

Reviewer: ```suggestion
Gardener autoscaler should be synced with respective [Kubernetes Autoscaler 1.x.y Release](https://github.com/kubernetes/autoscaler/releases/)
```

Reviewer: The title still has `v1.2x` . Kindly fix that.

Author: fixed.
